### PR TITLE
Add Bundler 2.4 blog post and release notes

### PR DESCRIPTION
**Pull request by deivid-rodriguez**
### What was the end-user problem that led to this PR?

The problem was the we should announce the upcoming Bundler 2.4 release.

### What is your fix for the problem, implemented in this PR?

My fix is to draft a release blog post and a "What's new" page.

### CODE REVIEW

File: source/blog/2022-12-22-bundler-v2-4.html.markdown [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+We took new year's release to move on and get rid of some stuff that was causing
+maintenance burden for us:
+
+* Ruby 2.3, 2.4, and 2.5 are no longer supported.
+* RubyGems 2.5, 2.6, and 2.7 are no longer supported.
+
+In general, this support drop should not break anything because RubyGems should
+be able to choose the latest supported Bundler on the Ruby version that you're
+using. But there's still some old RubyGems out there that don't have this
```
Comment (olleolleolle) [inline source/blog/2022-12-22-bundler-v2-4.html.markdown:152]:
```suggestion
using. But there are still some old RubyGems out there that don't have this
```

File: source/blog/2022-12-22-bundler-v2-4.html.markdown [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+feature, and `gem install bundler` command that could break there. We have
```
Comment (olleolleolle) [inline source/blog/2022-12-22-bundler-v2-4.html.markdown:153]:
```suggestion
feature, and the `gem install bundler` command could break there. We have
```

**Comment by deivid-rodriguez** [comment]
Thanks so much for copy editing! ❤️

File: source/blog/2022-12-22-bundler-v2-4.html.markdown [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+Now we just clone what's strictly necessary for Bundler to work, resulting in
+big disk space savings, and much faster bundling.
+
+### New CLI features
+
+We added a few small CLI features, such as,
+
+* `bundle lock --update --bundler`, to update your locked version of Bundler
+  without installing any gems.
+* A new `--pre` flat to `bundle update` and `bundle lock` to explicitly opt-in
```
[the remark this answers is not in the document]
Comment (deivid-rodriguez) [reply source/blog/2022-12-22-bundler-v2-4.html.markdown:138]:
```suggestion
* A new `--pre` flag to `bundle update` and `bundle lock` to explicitly opt-in
```

File: source/v2.4/whats_new.html.haml [hunk 4]
```diff
@@ -0,0 +1,49 @@
+.container.guide
+  %h1
+    What's New in
+    = current_visible_version
+
+  %p
+    The
+    = link_to 'Bundler 2.4 announcement', '/blog/2022/12/22/bundler-v2-4.html'
+    includes context and a more detailed explanation of the changes in this version. This is a summary of the biggest changes. As always, a detailed list of every change is provided in
+    == #{link_to "the changelog", "https://github.com/rubygems/rubygems/blob/3.4/bundler/CHANGELOG.md"}.
```
Comment (olleolleolle) [inline source/v2.4/whats_new.html.haml:10]:
```suggestion
    = link_to "the changelog", "https://github.com/rubygems/rubygems/blob/3.4/bundler/CHANGELOG.md".
```

Comment (olleolleolle) [reply source/v2.4/whats_new.html.haml:10]: Haml Lint complained:

> source/v2.4/whats_new.html.haml:10 [W] SpaceBeforeScript: The = symbol should have one space separating it from code
> source/v2.4/whats_new.html.haml:10 [W] UnnecessaryStringOutput: `= "..."` should be rewritten as `...`

**Review by olleolleolle — COMMENTED** [review]
Right, I found a small thing.

**Comment by deivid-rodriguez** [comment]
Blog post draft is here: https://bundler-site-bundler-2--j2zpxa.herokuapp.com/blog/2022/12/22/bundler-v2-4.html

I left the section about gems with Rust extensions empty because I think @simi wanted to write about this.

What's new page is here: https://bundler-site-bundler-2--j2zpxa.herokuapp.com/v2.4/whats_new.html.

**Comment by deivid-rodriguez** [comment]
I moved additional changes to other PRs to make reviewing this one easier.

**Comment by simi** [comment]
> Blog post draft is here: https://bundler-site-bundler-2--j2zpxa.herokuapp.com/blog/2022/12/22/bundler-v2-4.html
> 
> I left the section about gems with Rust extensions empty because I think @simi wanted to write about this.
> 
> What's new page is here: https://bundler-site-bundler-2--j2zpxa.herokuapp.com/v2.4/whats_new.html.

I have prepared initial version at https://github.com/rubygems/bundler-site/pull/982.

File: source/blog/2022-12-22-bundler-v2-4.html.markdown [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+  version solving has failed.
+~~~
+
+We tried to make this migration as backwards-compatible as possible, but
+there's a chance of experiencing some different solutions to the ones found by
+Molinillo, since the version solving problem does not have unique solutions.
+Please report any issues you find with the new resolver.
+
+
+### Easily generate gems with Rust extensions using `bundle gem`
```
Comment (deivid-rodriguez) [inline source/blog/2022-12-22-bundler-v2-4.html.markdown:107]: @simi So the idea would be to ditch this section and publish a separate blog post, correct?

Comment (simi) [reply source/blog/2022-12-22-bundler-v2-4.html.markdown:107]: I think you can keep it in here, just write a one small sentence with link or make whole title a link to the other post.

Comment (deivid-rodriguez) [reply source/blog/2022-12-22-bundler-v2-4.html.markdown:107]: Sounds good, let me update it!

Comment (deivid-rodriguez) [reply source/blog/2022-12-22-bundler-v2-4.html.markdown:107]: May I also include your blog post with this PR, so it's easier to review/merge everything at once?

Comment (deivid-rodriguez) [reply source/blog/2022-12-22-bundler-v2-4.html.markdown]: I just did that, and included a link to it here!

Comment (simi) [reply source/blog/2022-12-22-bundler-v2-4.html.markdown]: :muscle:

File: source/blog/2022-12-25-bundler-v2-4.html.markdown [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+several improvements that we hope will make our users happy :)
+
+## A better, PubGrub based, resolver
+
+Bundler now uses the most advanced algorithm to resolve versions, PubGrub.
+Kudos to Natalie Weizenbaum for [inventing
+it](https://nex3.medium.com/pubgrub-2fb6470504f) and to John Hawthorn for
+[porting it to Ruby](https://github.com/jhawthorn/pub_grub)!
+
+Our previous resolver, Molinillo, worked pretty well, but it really got in the
```
Comment (tnir) [inline source/blog/2022-12-25-bundler-v2-4.html.markdown:20]: I guess many readers (users of Bundler) do not know what Molinilo is. Why not linking the repo?

```suggestion
Our previous resolver, [Molinillo](https://github.com/CocoaPods/Molinillo), worked pretty well, but it really got in the
```

Or we can link to https://rubygems.org/gems/molinillo on `RubyGems.org`.

File: source/blog/2022-12-25-bundler-v2-4.html.markdown [hunk 7]
```diff
@@ -0,0 +1,170 @@
+---
+title: "Bundler v2.4: new resolver, gems with Rust extensions, and more"
+date: 2022-12-25 06:35 UTC
+tags:
+author: David Rodríguez
+author_url: https://github.com/deivid-rodriguez/
```
Comment (tnir) [inline source/blog/2022-12-25-bundler-v2-4.html.markdown:6]: One without the trailing slash looks more canonical to us:
```suggestion
author_url: https://github.com/deivid-rodriguez
```

File: source/blog/2022-12-25-rust-gem-skeleton.html.markdown [hunk 8]
```diff
@@ -0,0 +1,156 @@
+---
+title: "Generate gem skeleton with Rust extension"
+date: 2022-12-25 20:00 UTC
+tags:
+author: Josef Šimánek
+author_url: https://github.com/simi/
```
Comment (tnir) [inline source/blog/2022-12-25-rust-gem-skeleton.html.markdown:6]: One without the trailing slash looks more canonical to us:
```suggestion
author_url: https://github.com/simi
```

**Review by tnir — COMMENTED** [review]
Looks nice! Minor comments.

File: helpers/config_helper.rb [hunk 9]
```diff
@@ -14,9 +14,9 @@ def versions_grouped_by_status
   private
 
   def status(version)
-    if version == "v2.3"
+    if version == "v2.4"
       "Current release"
-    elsif %w[v2.2 v2.1].include?(version)
+    elsif %w[v2.3 2.2 v2.1].include?(version)
```
Comment (olleolleolle) [inline helpers/config_helper.rb:19]:
```suggestion
    elsif %w[v2.3 v2.2 v2.1].include?(version)
```

File: source/blog/2022-12-25-rust-gem-skeleton.html.markdown [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+~~~
+
+And finally, it is possible to call `hello` method defined in Rust returning string and print it to the console.
+
+~~~
+$ bundle exec ruby -rhello_rust -e 'puts HelloRust.hello("Josef")'
+"Hello from Rust, Josef!"
+~~~
+
+*Feel free to try to break this extension. For example you can try to pass different type of argument (like number or symbol). [magnus](https://github.com/matsadler/magnus) is doing great job [automatically coverting](https://github.com/matsadler/magnus#defining-methods) all those mistakes with friendly error messages.*
```
Comment (olleolleolle) [inline source/blog/2022-12-25-rust-gem-skeleton.html.markdown:150]:
```suggestion
*Feel free to try to break this extension. For example you can try to pass different types of argument (like number or symbol). [magnus](https://github.com/matsadler/magnus) is doing a great job [automatically converting](https://github.com/matsadler/magnus#defining-methods) all those mistakes with friendly error messages.*
```

File: source/blog/2022-12-25-rust-gem-skeleton.html.markdown [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+/usr/bin/gmake install target_prefix=
+generating target/release/libhello_rust.so (release)
+cargo rustc --target-dir target --manifest-path ../../../../ext/hello_rust/Cargo.toml --lib --release -- -C linker=gcc -L native=/home/retro/.rubies/ruby-3.1.2/lib -C link-arg=-lm
+    Finished release [optimized] target(s) in 0.09s
+installing hello_rust.so to /home/retro/code/hello_rust/lib/hello_rust
+/usr/bin/install -c -m 0755 hello_rust.so /home/retro/code//hello_rust/lib/hello_rust
+cp tmp/x86_64-linux/hello_rust/3.1.2/hello_rust.so tmp/x86_64-linux/stage/lib/hello_rust/hello_rust.so
+~~~
+
+And finally, it is possible to call `hello` method defined in Rust returning string and print it to the console.
```
Comment (olleolleolle) [inline source/blog/2022-12-25-rust-gem-skeleton.html.markdown:143]:
```suggestion
And finally, it is possible to call `hello` method defined in Rust returning a string and printing it to the console.
```

File: source/blog/2022-12-25-rust-gem-skeleton.html.markdown [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+Using minitest 5.16.3
+Using rake-compiler 1.2.1
+Using rb_sys 0.9.52
+Bundle complete! 5 Gemfile dependencies, 6 gems now installed.
+Use `bundle info [gemname]` to see where a bundled gem is installed.
+~~~
+
+At this stage, everything is ready to compile Rust code and glue it with Ruby.
+
+*You need to have Rust already installed on your system. See [rustup](https://rustup.rs/) for simple installation experience.*
```
Comment (olleolleolle) [inline source/blog/2022-12-25-rust-gem-skeleton.html.markdown:113]:
```suggestion
*You need to have Rust already installed on your system. See [rustup](https://rustup.rs/) for a simple installation experience.*
```

File: helpers/config_helper.rb (hunk 9 above)
Comment (deivid-rodriguez) [reply helpers/config_helper.rb:19]: :O Awesome catch!

**Review by simi — APPROVED** [review]

**Comment by tnir** [comment]
@deivid-rodriguez What is the blocker for merging this?

**Comment by deivid-rodriguez** [comment]
I'm finishing up a fix to `bundle lock --update --bundler` command, which is currently advertised in the blog post.

I want it to work fine before we publish the post, but if we are willing to publish this right now, I can edit the blog post to remove the mention to it and publish it.

**Comment by tnir** [comment]
Thanks @deivid-rodriguez. If you do not want to publish this blog post, I am still inclined to publish Bundler v2.4-based docs (`https://bundler.io/v2.4/docs.html`) as soon as possible. To do so, we can create another PR from this but without changes in `/source/blog`. What do you think?

**Comment by deivid-rodriguez** [comment]
Sounds good, let me extract a separate PR!

**Comment by deivid-rodriguez** [comment]
I extracted https://github.com/rubygems/bundler-site/pull/991.

**Comment by tnir** [comment]
#1051 might be relevant to this once it is merged.

File: source/blog/2022-12-25-rust-gem-skeleton.html.markdown [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+
+And now, with Bundler 2.4, `bundle gem` skeleton generator can provide all the glue you need to start using Rust inside your gems thanks to the new `--ext=rust` parameter!
+
+## What's new?
+
+Thanks to new parameter it is possible to generate simple Rust-based gem extension.
+
+*Notice I already have bundle `gem command` configured. Your output can differ. When running `bundle gem` for the first time, it will interactively ask you few questions.*
```
Comment (simi) [inline source/blog/2022-12-25-rust-gem-skeleton.html.markdown:20]: Should we recommend to update to at least `RubyGems 3.4.6` here now?

Comment (deivid-rodriguez) [reply source/blog/2022-12-25-rust-gem-skeleton.html.markdown:20]: Yeah! Let me update this!

**Review by simi — APPROVED** [review]

**Comment by deivid-rodriguez** [comment]
Preview looks good, so let's publish this!

**Pull request merged**